### DISCUSSION TITLE
File Headers

### DISCUSSION THREAD

**Discussion by zcwilt** [zencart/zencart#5216] [General]
Zen Cart currently includes a set of file headers

e.g.

```
/**
 * @copyright Copyright 2003-2022 Zen Cart Development Team
 * @copyright Portions Copyright 2003 osCommerce
 * @license http://www.zen-cart.com/license/2_0.txt GNU Public License V2.0
 * @version $Id: DrByte 2020 Dec 25 Modified in v1.5.8-alpha $
 */
```

Some of those headers are static, while others are updated at release/pre-release with information based on the commits made to the file.

Some history here.

Originally Zen Cart hosted its code on source forge and used CVS to manage version history. This was private access and only founders had access to creating commits. At that time we only had minimal headers (basically just copyright headers)

Later on we moved to using a private SVN repository as SVN gave us better control over branching.

This was still private to Zen Cart founders.

SVN allowed for keyword expansion in the headers.

E.g. by just adding @version $Id it would add the commiter name and commit date.

Some people found this useful in diffing between releases when doing upgrades.

(again based on the fact that our svn repository was only available to founders)

When we moved to using git (which doesn't do keyword expansion) we started to build the versionstamper code
that updates the headers. The main extra thing this added was the notification of where code had changed 
e.g Modified in v1.5.8-alpha $ 

Again this was before we gave full access to version history through github. 

Given that developers now have full access to version history through git/github there is an argument that we no longer need these headers and don't need the versionstamper to be run on releases.

e.g. we can git blame to see authors and diffing tools can show where a file was created/modified

Legally as well updating copyright headers every year isn't really needed.

The question is - are people using the headers we have now to manage there upgrades from version to version
Should we still version stamp or move to purely static headers.

**Comment by zcwilt** [comment]
Just to clarify

Not saying we don't need headers. Just that those headers will be static and not updated based on commit history

**Comment by scottcwilson** [comment]
As a simplification, I would be in favor of removing the author name.  The version info is useful in determining what the baseline version of the file was (example: 1.5.7b) to see if changes have been made.

**Comment by dbltoe** [comment]
I will just follow as directed but I have some observations to point out. 

The blame button (I don't care for the term) does not seem to work for the header portion.  One of the recent files I looked at with "blame" showed that torvista had made a change in the body but did not flag the change of version id that changed it to torvista and 1.5.8.

I use the version id extensively in comparing files from upgrades, security fixes, and mod file changes.  Thus the reason I added my changes to the files I submitted.  Since alpha 2 was "active", I included that.

There is a tremendous advantage to seeing when a file was changed.  I think who made the change should be included, not to gain notoriety but to know who to contact for questions.  Not everyone working on a site would even know to go to git and navigate to the page that includes the "blame" button.

edit:  If the author is kept, I would suggest the use of the author's forum username thus adding even more ability to discuss.

**Reply by scottcwilson** [reply]
>  I think who made the change should be included ... to know who to contact for questions.

Yeah don't do that.  This is EXTREMELY annoying.  Post questions to the forum.

**Comment by simon1066** [comment]
I too use the version_id, albeit occassionally, during ZC upgrade - to identify the more complicated file changes made by some mods.

**Reply by scottcwilson** [reply]
Exactly.

**Comment by barco57** [comment]
I use the header information mostly when fixing an upgrade that was not done well to determine what version a file is when it didn't get updated to ferret out better who to "Blame" for it by tracking the upgrade history of the site. Blame is the correct word as I am not politically correct...

**Reply by barco57** [reply]
I don't so much worry about the author as I am normally not blaming the author of the file, I am looking to blame the person who did the upgrade and left an old file behind and what base version of ZC the file comes from to line up with the current upgrade or previous upgrades where it has been left behind multiple times for some reason to build my body of evidence to slap someone around, which I do take pleasure in... never argue with the guy who has the server logs

**Reply by scottcwilson** [reply]
... and sometimes plugins use old versions of files, so you really have to be sure of your baseline and diff against that.

**Reply by barco57** [reply]
true, and sometimes the mods code in that older file is not well commented

**Comment by zcwilt** [comment]
Thanks for the feedback. 

I would note a couple of things. 

The author entry in the id doesn't always attribute the author mainly responsible.

I could write a whole new class and then maybe some one spots a very minor change needed, commits and they would be the attributed author.

Again version stamping of the header will only show the most recent committer not the original author.

Author names are taken from whatever the was set for the config author, generally this will be their github handle, but not always, we have no control of that (except in some limited circumstances) 

I think I will probably remove the update of copyright date. this is not legally required and reduces the surface area of changes.
We will continue to update the ID line for now.

Some documentation should be produced to show what an initial header looks like. 
e.g there are implications of adding additional copyright lines 
for example 
` * @copyright Portions Copyright 2003 osCommerce`

**Comment by torvista** [comment]
Can't you just use the last-committed date in the file header to identify when it was last modified?
No author name, no to-which-fileset-it-belongs, no filesize, no version, no id. The file date can be the id.
Once you see the date is modified from your old site file, you have to compare the whole file anyway (if you have modded it), all you need is one indication of a change to give you the heads-up.
If anyone wants history it's all here on Github. That's the point of Github no?

**Comment by scottcwilson** [comment]
> Can't you just use the last-committed date in the file header to identify when it was last modified?

No.  I need a fast way of locating a file with an unknown provenance, and the version is the easiest way to do it. 
A file date does not allow me to deduce that the file came from 1.5.6c, not 1.5.6f, which is the kind of thing I have to do all the time.
